core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -225,15 +253,52 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
+            cancelScheduleTask();
+            requestConnection(); // is recursive so might hit the end of the addresses
+          }
 
-          // If no addresses remaining, go into TRANSIENT_FAILURE
           if (!addressIndex.isValid()) {
+            // If no addresses remaining, go into TRANSIENT_FAILURE
             helper.refreshNameResolution();
             rawConnectivityState = TRANSIENT_FAILURE;
             updateBalancingState(TRANSIENT_FAILURE,
                 new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+        } else if (!addressIndex.isValid()) {
+          if (concludedState == TRANSIENT_FAILURE) {
+            // sticky TRANSIENT_FAILURE
+            helper.refreshNameResolution();
+            updateBalancingState(TRANSIENT_FAILURE,
+                new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+
+          boolean hasIdle = false;
```

Comment:
I might miss something, could you point me to the section in grfc about the hasIdle/hasConnecting thing.
Also what is missing here is to count the TF times starting from the first pass based on grfc `Happy Eyeballs in the pick_first LB Policy
` section

> If the first pass completes without a successful connection attempt, ... We will count the number of connection failures, and when that number reaches the number of subchannels, we will request re-resolution;

Reply:
Updated to only do the refreshNameResolution when there have been sufficient connection failures.

Reply:
Happy Eyeballs should be the same as the petiole policies which had the rule that if any child is READY, then it is READY, if not but any child is CONNECTING then it is CONNECTING, if not but some child is IDLE then it is IDLE, otherwise it is TRANSIENT FAILURE.

Reply:
Happy eyeballs is pick first not petiole policies. I don't think aggregated connectivity applies here this way. The aggregated connectivity processing here makes it unnecessarily complicated.